Confidential Ledger samples (#20930)
Azure/azure-sdk-for-net: sdk/confidentialledger/Azure.Storage.ConfidentialLedger/README.md

#### Using Azure Active Directory

> This document demonstrates using [DefaultAzureCredential][default_cred_ref] to authenticate to the Confidential Ledger via Azure Active Directory. However, `ConfidentialLedgerClient` accepts any [Azure.Identity][azure_identity] token credential. See the [Azure.Identity][azure_identity] documentation for more information about other credentials.

Reviewer: ```suggestion
This document demonstrates using [DefaultAzureCredential][default_cred_ref] to authenticate to the Confidential Ledger via Azure Active Directory. However, any of the credentials offered by the [Azure.Identity][azure_identity] will be accepted.  See the [Azure.Identity][azure_identity] documentation for more information about other credentials.
```

```
Then, `DefaultAzureCredential` will be able to authenticate the `ConfidentialLedgerClient`.

> Constructing the client also requires your Confidential Ledger's URL and id, which you can get from the Azure CLI or the Azure Portal.  When you have retrieved those values, please replace instances of `"my-ledger-id"` and `"https://my-ledger-url.confidential-ledger.azure.com"` in the examples below

Reviewer: Should we include an example of the CLI or a link to documentation for how to use it to retrieve the values?

Response response = identityClient.GetLedgerIdentity(ledgerId);

// extract the ECC PEM value from the response.
var eccPem = JsonDocument.Parse(response.Content)
    .RootElement
    .GetProperty("ledgerTlsCertificate")
    .GetString();

// construct an X509Certificate2 with the ECC PEM value.
X509Certificate2 ledgerTlsCert = new X509Certificate2(Encoding.UTF8.GetBytes(eccPem));
```

Now we can construct the `ConfidentialLedgerClient` with a transport configuration that trusts the `ledgerTlsCert`.

```C# Snippet:CreateClient
// Create a certificate chain rooted with our TLS cert.
X509Chain certificateChain = new();
certificateChain.ChainPolicy.RevocationMode = X509RevocationMode.NoCheck;
certificateChain.ChainPolicy.RevocationFlag = X509RevocationFlag.ExcludeRoot;
certificateChain.ChainPolicy.VerificationFlags = X509VerificationFlags.AllowUnknownCertificateAuthority;
certificateChain.ChainPolicy.VerificationTime = DateTime.Now;
certificateChain.ChainPolicy.UrlRetrievalTimeout = new TimeSpan(0, 0, 0);
certificateChain.ChainPolicy.ExtraStore.Add(ledgerTlsCert);

// Define a validation function to ensure that the ledger certificate is trusted by the ledger identity TLS certificate.

> Func<HttpRequestMessage, X509Certificate2, X509Chain, SslPolicyErrors, bool> certValidationCheck;

Reviewer: Should we consider using a local function instead of the `Func` definition?  That can often reduce visual clutter and be easier to read.

Author: I went back and forth on this, so I'll lean in your suggested direction.

Console.WriteLine(contents); // "Hello world!"

// Now just provide the transactionId.
getBySubledgerResponse = ledgerClient.GetLedgerEntry(transactionId);

string subLedgerId2 = JsonDocument.Parse(getBySubledgerResponse.Content)
    .RootElement
    .GetProperty("subLedgerId")
    .GetString();

Console.WriteLine($"{subLedgerId} == {subLedgerId2}");
```

Ledger entries are retrieved from sub-ledgers. When a transaction id is specified, the returned value is the value contained in the specified sub-ledger at the point in time identified by the transaction id. If no transaction id is specified, the latest available value is returned.

```C# Snippet:GetEnteryWithNoTransactionId
Response firstPostResponse = ledgerClient.PostLedgerEntry(
    RequestContent.Create(new { contents = "Hello world 0" }));
ledgerClient.PostLedgerEntry(
    RequestContent.Create(new { contents = "Hello world 1" }));
Response subLedgerPostResponse = ledgerClient.PostLedgerEntry(
    RequestContent.Create(new { contents = "Hello world sub-ledger 0" }),
    "my sub-ledger");
ledgerClient.PostLedgerEntry(
    RequestContent.Create(new { contents = "Hello world sub-ledger 1" }),
    "my sub-ledger");

> firstPostResponse.Headers.TryGetValue("x-ms-ccf-transaction-id", out string transactionId);

Reviewer: `x-ms-ccf-transaction-id` gets used enough in the examples that I wonder if we should consider exposing it via some kind of `ConfidentialLedgerHeaders` constants class.
